language/move-prover/bytecode/src/global_invariant_instrumentation.rs

```diff
@@ -0,0 +1,497 @@
+    /// Once reaching a fixedpoint, we know that the `by_function_instantiation` captures all
+    /// invariants and their instantiations that are needed in that function instantiation.
+    fn self_join(self) -> Self {
+        let Self {
+            generic_condensation,
+            mut by_function_instantiation,
+        } = self;
+
+        let mut changed = true;
+        while changed {
+            changed = false;
+            let fun_insts: BTreeSet<_> = by_function_instantiation.keys().cloned().collect();
+
+            for lhs_inst in fun_insts {
+                let mut lhs_pack = by_function_instantiation.remove(&lhs_inst).unwrap();
+                for (rhs_inst, rhs_pack) in &by_function_instantiation {
+                    let adapter = TypeUnificationAdapter::new_vec(&lhs_inst, rhs_inst, false, true);
+                    if let Some((_, mut subst_rhs)) = adapter.unify(Variance::Allow, false) {
+                        let inst: Vec<_> = (0..rhs_inst.len() as u16)
+                            .map(|i| subst_rhs.remove(&i).unwrap_or(Type::TypeParameter(i)))
+                            .collect();
+                        if lhs_pack.join_pack(rhs_pack, &inst) {
+                            changed = true;
+                        }
+                    }
+                }
+                by_function_instantiation.insert(lhs_inst, lhs_pack);
+            }
+        }
+
+        Self {
+            generic_condensation,
+            by_function_instantiation,
+        }
+    }
+}
+
+// The function target processor
+pub struct GlobalInvariantInstrumentationProcessor {}
+
+impl GlobalInvariantInstrumentationProcessor {
+    pub fn new() -> Box<Self> {
+        Box::new(Self {})
+    }
+}
+
+impl FunctionTargetProcessor for GlobalInvariantInstrumentationProcessor {
+    fn process(
+        &self,
+        targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv<'_>,
+        data: FunctionData,
+    ) -> FunctionData {
+        if fun_env.is_native() || fun_env.is_intrinsic() {
+            // nothing to do
+            return data;
+        }
+        if !data.variant.is_verified() {
+            // only need to instrument if this is a verification variant
+            return data;
+        }
+        debug_assert!(matches!(
+            data.variant,
+            FunctionVariant::Verification(VerificationFlavor::Regular)
+        ));
+
+        let env = fun_env.module_env.env;
+        let options = ProverOptions::get(env);
+
+        // retrieve and transpose the analysis result
+        let target = FunctionTarget::new(fun_env, &data);
+        let analysis_result = global_invariant_analysis::get_info(&target);
+        let summary = InstrumentationSummary::transpose(analysis_result);
+
+        // if the backend supports some form of monomorphization, instrument the invariants in
+        // the generic version of the function only.
+        if options.boogie_poly {
```

Comment:
I thought that by default we only instrument generic functions, and then instantiate them via monomorphization (and in the Boogie backend, which has all the logic for instantiation already).

Reply:
Yes, I was trying to addressing this in the PR description as well.

> The instrumenter supports two mode of operations, depending on whether
> the prover backend supports monomorphization or not:
>
> - With the option --boogie-poly, the instrumenter will instrument
> instantiated invariants in the generic function (and the generic
> instance is the only function instance).
>
> - Without the --boogie-poly option, the instrumenter will instrument
> instantiated invariants per each instantiation of a generic
> function (this is the traditional workflow). And this means that a
> function will have multiple instances for verification.

I also listed some reasons why we had an option 2 here.

Reply:
Eventually we would want option 1: by default we only instrument generic functions, and then instantiate them via monomorphization. And I am currently implementing it. And this means that we might not even need this `--boogie-poly` flag here and remove the `else` branch of this `if options.boogie_poly`.

The main reason for this solution is to get at least the whole Diem Framework verifying again to test out the whole transformation pipeline, in order to boost our confidence.